### PR TITLE
Algo order resume permission

### PR DESCRIPTION
**Pull request by avsek477**
This feature gives users the ability to load/resume the selected algo orders in case the dms feature is disabled. Having the dms feature disabled means that the server doesn't cancel the active orders automatically on HF restart/close.

Proposed solution: UI call the server endpoint `get.active_algo_orders` and then shows a popup to the user with the given array of algo orders as received in response event `algo.active_orders`. Now, the user has the ability to resume the orders with the previously saved settings and values or, the user can remove those orders. Based on user's selection, UI then sends the selected orders in an array to the HF server for them to be either cancelled or resumed.

Also, in case of websocket reopen/reconnect, the server removes all the running instances of active running orders and emits an event called `algo.cancel_orders`. This is an indication to the UI to remove all the orders from the algo tab and then call the endpoint `get.active_algo_orders`  to fetch the active orders and show them in a popup for users to resume / cancel on demand.

Note:
1. For fetching active algo orders, 
 ```
dispatch(WSActions.send([
  'get.active_algo_orders',
  authToken,
]))
```
Server then fires a `algo.active_orders` event. Example Response:
`["algo.active_orders",[{ "gid": "1611908852446", "args": {}, "label": "", "name": ""}]]`

2. For cancelling previously active orders, send the desired orders array in the following format:
```
const orders = [
  { "gid": "1611908852446", "algoID": "bfx-twap" }
]
dispatch(WSActions.send([
  'algo_order.remove',
  authToken,
  orders,
]))
```
Upon successful order cancellation, an event 'algo.orders_removed' is fired with gid of removed orders
```
["algo.orders_removed", ["1612421103100"]]
```
3. For resuming/loading previously active orders, send the desired orders array in the following format:
```
const orders = [
  { "gid": "1611908852447", "algoID": "bfx-twap" }
]
dispatch(WSActions.send([
  'algo_order.load',
  authToken,
  orders
]))
```
Upon successful order reload, an event 'algo.order_loaded' is fired with gid of resumed order
```
["algo.order.loaded", "1612421103100"]
```
Related PR: https://github.com/bitfinexcom/bfx-hf-algo/pull/54

### CODE REVIEW

File: lib/ws_servers/api/index.js [hunk 1]
```diff
@@ -92,20 +98,48 @@ module.exports = class APIWSServer extends WSServer {
...
   }
 
+  async sendActiveAlgoOrders (ws) {
+    const { UserSettings } = this.db
+    const { AlgoOrder } = this.algoDB
+    const { userSettings: settings } = await UserSettings.getAll()
+    const { dms } = settings || DEFAULT_SETTINGS
+
+    if (dms) {
+      this.as.open()
```
Comment (robertkowalski) [inline lib/ws_servers/api/index.js:123]: afaik we should also ask the user, if the dms option is not set, just to make sure

File: lib/ws_servers/api/index.js [hunk 2]
```diff
@@ -92,20 +98,48 @@ module.exports = class APIWSServer extends WSServer {
...
+      return
+    }
+
+    const aos = await AlgoOrder.find([['active', '=', true]])
+
+    if (aos.length === 0) {
+      this.as.open()
+    } else {
```
Comment (robertkowalski) [inline lib/ws_servers/api/index.js:131]: you can reduce nesting here:

```js
    if (aos.length === 0) {
      this.as.open()
      return
    }

```

File: lib/ws_servers/api/index.js [hunk 3]
```diff
@@ -92,20 +98,48 @@ module.exports = class APIWSServer extends WSServer {
...
 
-    await PI.forEach(EXCHANGE_ADAPTERS, async ({ id }) => {
-      const { Market } = this.db
-      const markets = await Market.find([['exchange', '=', id]])
-
-      send(ws, ['info.markets', id, markets])
-    }).catch((err) => {
-      this.d('error sending markets to client: %s', err.stack)
-    })
+    for (const { id } of EXCHANGE_ADAPTERS) {
```
Comment (robertkowalski) [inline lib/ws_servers/api/index.js:101]: this is not part of the algo orders, can you revert?

Comment (robertkowalski) [reply lib/ws_servers/api/index.js:101]: i had to touch it for the paper trading feature, so its going to get refactored soon

File: lib/ws_servers/api/send_authenticated.js [hunk 4]
```diff
@@ -28,7 +27,7 @@ module.exports = async (ws, db, d, opts) => {
     .filter(c => c.cid !== CREDENTIALS_CID)
 
   // Attempt to decrypt & open connections for all valid credentials
-  await PI.forEach(apiCredentials, async (credentials) => {
+  for (const credentials of apiCredentials) {
```
Comment (robertkowalski) [inline lib/ws_servers/api/send_authenticated.js:30]: very nice!!

in general this belongs in a separate PR, has not much to do with algo orders?

Comment (robertkowalski) [reply lib/ws_servers/api/send_authenticated.js:30]: PS: as mentioned in slack i had to extend and change this section as part of the paper trading mode feature.

see my chat message that i refactor pi-iteration out and this PR

File: lib/ws_servers/api/send_strategies.js [hunk 5]
```diff
@@ -22,7 +21,7 @@ module.exports = async (ws, db, d) => {
 
   const decryptedStrategies = []
 
-  await PI.forEach(Object.values(strategiesByID), async (encryptedStrategy) => {
+  for (const encryptedStrategy of Object.values(strategiesByID)) {
```
Comment (robertkowalski) [inline lib/ws_servers/api/send_strategies.js:24]: can you put in a separate PR?

File: lib/ws_servers/api/send_authenticated.js (hunk 4 above)
Comment (robertkowalski) [reply lib/ws_servers/api/send_authenticated.js:30]: can you revert?

File: package.json [hunk 6]
```diff
@@ -53,5 +51,13 @@
...
     "unit": "NODE_PATH=lib NODE_ENV=test mocha"
   },
-  "devDependencies": {}
+  "devDependencies": {
+    "mocha": "^6.2.3",
+    "standard": "^16.0.3"
+  },
+  "standard": {
+    "ignore": [
+      "/docs/**/*.js"
```
Comment (robertkowalski) [inline package.json:60]: nice!

File: package.json [hunk 7]
```diff
@@ -53,5 +51,13 @@
     "test": "npm run lint && npm run unit",
     "unit": "NODE_PATH=lib NODE_ENV=test mocha"
   },
-  "devDependencies": {}
+  "devDependencies": {
+    "mocha": "^6.2.3",
```
Comment (robertkowalski) [inline package.json:55]: good catch!

File: lib/ws_servers/api/index.js (hunk 3 above)
Comment (avsek477) [reply lib/ws_servers/api/index.js:101]: will do

File: lib/ws_servers/api/send_authenticated.js (hunk 4 above)
Comment (avsek477) [reply lib/ws_servers/api/send_authenticated.js:30]: yeah .. i had this change before i saw your review .. so forgot .. i will revert this

File: lib/ws_servers/api/send_strategies.js (hunk 5 above)
Comment (avsek477) [reply lib/ws_servers/api/send_strategies.js:24]: will do ..

File: lib/ws_servers/api/handlers/on_algo_order_remove.js [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+  })
+
+  if (!validRequest) {
+    return
+  }
+
+  if (!isAuthorized(ws, authToken)) {
+    return sendError(ws, 'Unauthorized')
+  } else if (exID !== 'bitfinex') {
+    return sendError(ws, 'Algo orders currently only enabled for Bitfinex')
```
Comment (robertkowalski) [inline lib/ws_servers/api/handlers/on_algo_order_remove.js:25]: we only support bitfinex, you can remove

File: lib/ws_servers/api/handlers/on_algo_order_remove.js [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+  const [, authToken, exID, orders] = msg
+
+  const validRequest = validateParams(ws, {
+    exID: { type: 'string', v: exID },
+    authToken: { type: 'string', v: authToken },
+    orders: { type: 'array', v: orders }
+  })
+
+  if (!validRequest) {
+    return
```
Comment (robertkowalski) [inline lib/ws_servers/api/handlers/on_algo_order_remove.js:19]: `d('invalid request: algo:remove')` makes debugging easier

File: lib/ws_servers/api/handlers/on_algo_order_remove.js [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    return sendError(ws, 'Unauthorized')
+  }
+
+  for (const algOrder of orders) {
+    const { gid, algoID } = algOrder
+    const { AlgoOrder } = algoDB
+    await AlgoOrder.update({ gid, algoID }, { active: false })
+  }
+
+  as.open()
```
Comment (robertkowalski) [inline lib/ws_servers/api/handlers/on_algo_order_remove.js:36]: currently `as.open()` is called in these cases:

 - no algo order found
 - algo orders were removed
 - dms = false

am i missing something or is there a case missing were someone does not want to remove any orders and the aos is not started? afaik the current implementation requires a request to "remove algo orders" with an empty array in those cases?

Comment (robertkowalski) [reply lib/ws_servers/api/handlers/on_algo_order_remove.js:36]: how about extending the PR a bit:

1. the algo order host is always started 
2. db is read from the outside, then algo orders that are supposed to start are sent to the algo host
3. the remove endpoint just removes algo orders from the db
4. on start of the algo server, no algo orders are started automatically (i.e. you don't pass active algo orders into the constructor)

**Review by robertkowalski — CHANGES_REQUESTED** [review]
nice work!

maybe we can improve the current algo order host even more

File: index.js [hunk 11]
```diff
@@ -70,7 +76,7 @@ module.exports = ({
   const opts = { wsURL: bfxWSURL, restURL: bfxRestURL }
   syncMarkets(apiDB, EXAS, opts).then(() => {
     as.open()
-
+
```
Comment (robertkowalski) [inline index.js:79]: additional whitespace?

File: lib/ws_servers/algos/handlers/on_open.js [hunk 12]
```diff
@@ -62,6 +63,21 @@ module.exports = async (server, ws, msg) => {
...
+    const aos = await AlgoOrder.find([['active', '=', true]])
+    
+    if (aos.length === 0) {
+      return
+    }
+
+    // aoc.userID = userID
+    // await aoc.loadAOs('bitfinex', aos)
+
+    // d('reloaded all active aos: %s', JSON.stringify(aos.map(ao => ao.gid)))
```
Comment (robertkowalski) [inline lib/ws_servers/algos/handlers/on_open.js:78]: is this finished?

File: lib/ws_servers/algos/handlers/on_open.js [hunk 13]
```diff
@@ -70,5 +86,11 @@ module.exports = async (server, ws, msg) => {
     server.broadcast(userID, ['stopped', userID, exID, gid])
   })
 
+  hosts[key].on('ao:persist:db:update', async(updateOpts) => {
```
Comment (robertkowalski) [inline lib/ws_servers/algos/handlers/on_open.js:89]: style

```js
hosts[key].on('ao:persist:db:update', async (updateOpts) => {
```

File: lib/ws_servers/algos/spawn_bitfinex_ao_host.js [hunk 14]
```diff
@@ -1,23 +1,18 @@
...
 
 const { _default: DEFAULT_SETTINGS } = require('bfx-hf-ui-config').UserSettings
-const HFDB = require('bfx-hf-models')
-const HFDBLowDBAdapter = require('bfx-hf-models-adapter-lowdb')
-const {
-  schema: HFDBBitfinexSchema
-} = require('bfx-hf-ext-plugin-bitfinex')
 
-const { AOHost } = require('bfx-hf-algo')
+const { AOHost } = require('../../../../bfx-hf-algo')
```
Comment (robertkowalski) [inline lib/ws_servers/algos/spawn_bitfinex_ao_host.js:5]: is this a leftover?

do you know npm link?

```
npm link bfx-hf-algo
```

File: lib/ws_servers/algos/spawn_bitfinex_ao_host.js [hunk 15]
```diff
@@ -1,23 +1,18 @@
...
 const {
   PingPong, Iceberg, TWAP, AccumulateDistribute, MACrossover, OCOCO
-} = require('bfx-hf-algo')
+} = require('../../../../bfx-hf-algo')
```
Comment (robertkowalski) [inline lib/ws_servers/algos/spawn_bitfinex_ao_host.js:8]: same here

File: lib/ws_servers/api/handlers/on_algo_order_remove.js [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+    return sendError(ws, 'Unauthorized')
+  }
+
+  for (const algOrder of orders) {
+    const { gid, algoID } = algOrder
+    const { AlgoOrder } = algoDB
+    await AlgoOrder.update({ gid, algoID }, { active: false })
+  }
+
+  d('removed all selected orders %s', JSON.stringify(orders.map(o => o.gid)))
```
Comment (robertkowalski) [inline lib/ws_servers/api/handlers/on_algo_order_remove.js:35]: no confirmation for the client?

would expect somethign like a:

`send('algo.orders_removed', $successful-ids)`

(pseudo-code)

File: lib/ws_servers/api/index.js [hunk 17]
```diff
@@ -106,6 +111,31 @@ module.exports = class APIWSServer extends WSServer {
...
   }
 
+  async sendActiveAlgoOrders (ws) {
+    const { UserSettings } = this.db
+    const { AlgoOrder } = this.algoDB
+    const { userSettings: settings = {} } = await UserSettings.getAll()
+    const { dms } = settings
+
+    if (dms) {
+      return
```
Comment (robertkowalski) [inline lib/ws_servers/api/index.js:121]: why cancelling the feature when DMS is enabled?

Comment (robertkowalski) [reply lib/ws_servers/api/index.js:121]: (it shoudl always work)

Comment (robertkowalski) [reply lib/ws_servers/api/index.js:121]: DMS cancels open orders, but not algo orders

File: lib/ws_servers/api/index.js [hunk 18]
```diff
@@ -106,6 +111,31 @@ module.exports = class APIWSServer extends WSServer {
...
+    }
+
+    const aos = await AlgoOrder.find([['active', '=', true]])
+
+    if (aos.length === 0) {
+      return
```
Comment (robertkowalski) [inline lib/ws_servers/api/index.js:127]: why not sending an empty array here with `info.active_algos `

Comment (robertkowalski) [reply lib/ws_servers/api/index.js:127]: (UI dev becomes a bit easier)

File: lib/ws_servers/api/handlers/on_algo_order_load.js [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+    return
+  }
+
+  if (!isAuthorized(ws, authToken)) {
+    return sendError(ws, 'Unauthorized')
+  } else if (!ws.aoc) {
+    return sendError(ws, 'Unauthorized')
+  }
+
+  ws.aoc.loadAOs(exID, orders)
```
Comment (robertkowalski) [inline lib/ws_servers/api/handlers/on_algo_order_load.js:28]: confirmation for the client?

File: lib/ws_servers/algos/index.js [hunk 20]
```diff
@@ -39,9 +43,11 @@ module.exports = class AlgoServer extends WSServer {
     this.wsURL = wsURL
     this.restURL = restURL
     this.apiDB = apiDB
+    this.algoDB = algoDB
     this.dbPath = hfLowDBPath
     this.clients = {}
     this.hosts = {}
+    this.aoc = new AlgoServerClient({ url: `http://localhost:${port}` })
```
Comment (robertkowalski) [inline lib/ws_servers/algos/index.js:50]: wait, is this required? the algo orders server sets up a client? why a client-server model then?

File: lib/ws_servers/algos/handlers/on_load.js [hunk 21]
```diff
@@ -0,0 +1,52 @@
...
+    try {
+      const ao = await AlgoOrder.get({ algoID, gid })
+      let { state } = ao || {}
+      if (!state) {
+        continue
+      }
+      state = JSON.parse(state)
+      state.active = true
+      await host.loadAO(algoID, gid, state)
+      d('AO loaded for user %s on exchange %s [%s]', userID, exID, gid)
```
Comment (robertkowalski) [inline lib/ws_servers/algos/handlers/on_load.js:46]: reply to client?

File: lib/ws_servers/algos/handlers/on_open.js [hunk 22]
```diff
@@ -62,6 +63,21 @@ module.exports = async (server, ws, msg) => {
     server.broadcast(userID, ['started', userID, exID, name, label, gid, args])
   })
 
+  hosts[key].on('meta:reload', async() => {
```
Comment (robertkowalski) [inline lib/ws_servers/algos/handlers/on_open.js:66]: style, missing space

**Review by robertkowalski — CHANGES_REQUESTED** [review]
good progress so far, some issues found

File: lib/ws_servers/algos/handlers/on_open.js (hunk 12 above)
Comment (robertkowalski) [reply lib/ws_servers/algos/handlers/on_open.js:78]: forget my question, just saw your commetn on slack :)

File: lib/ws_servers/api/handlers/on_algo_order_remove.js [hunk 23]
```diff
@@ -0,0 +1,36 @@
...
+  if (!isAuthorized(ws, authToken)) {
+    return sendError(ws, 'Unauthorized')
+  } else if (!ws.aoc) {
+    return sendError(ws, 'Unauthorized')
+  }
+
+  for (const algOrder of orders) {
+    const { gid, algoID } = algOrder
+    const { AlgoOrder } = algoDB
+    await AlgoOrder.update({ gid, algoID }, { active: false })
```
Comment (robertkowalski) [inline lib/ws_servers/api/handlers/on_algo_order_remove.js:32]: does it work on runtime? 

should we stop the algo order first on the algo host?

File: lib/ws_servers/algos/handlers/on_open.js [hunk 24]
```diff
@@ -62,6 +63,21 @@ module.exports = async (server, ws, msg) => {
...
+    const { AlgoOrder } = algoDB
+
+    const aos = await AlgoOrder.find([['active', '=', true]])
+    
+    if (aos.length === 0) {
+      return
+    }
+
+    // aoc.userID = userID
+    // await aoc.loadAOs('bitfinex', aos)
```
Comment (robertkowalski) [inline lib/ws_servers/algos/handlers/on_open.js:76]: hm, server takes the client and calls itself... not a good idea...

Comment (robertkowalski) [reply lib/ws_servers/algos/handlers/on_open.js:76]: how about broadcasting to the algo client, like in other cases too. 

another layer of indirection we might want to delete in the future, but follows the client/server model

Comment (robertkowalski) [reply lib/ws_servers/algos/handlers/on_open.js:76]: 1. create a handler in ws_clients/algos, e.g. for the rpc command "foo"
2. broadcast: `server.broadcast(userID, ['foo', userID])`
3. in the algo orders client, find active orders and run the command

**Review by robertkowalski — APPROVED** [review]
lgtm
